src/test/operators/join_test_runner.cpp

```diff
@@ -0,0 +1,527 @@
+  InputSide side{};
+  ChunkOffset chunk_size{};
+  size_t table_size{};
+  InputTableType table_type{};
+
+  auto to_tuple() const { return std::tie(side, chunk_size, table_size, table_type); }
+};
+
+bool operator<(const InputTableConfiguration& l, const InputTableConfiguration& r) {
+  return l.to_tuple() < r.to_tuple();
+}
+
+class BaseJoinOperatorFactory;
+
+struct JoinTestConfiguration {
+  InputTableConfiguration input_left;
+  InputTableConfiguration input_right;
+  JoinMode join_mode{JoinMode::Inner};
+  DataType data_type_left{DataType::Int};
+  DataType data_type_right{DataType::Int};
+  bool nullable_left{false};
+  bool nullable_right{false};
+  PredicateCondition predicate_condition{PredicateCondition::Equals};
+  std::vector<OperatorJoinPredicate> secondary_predicates;
+  std::shared_ptr<BaseJoinOperatorFactory> join_operator_factory;
+
+  // Only for JoinHash
+  std::optional<size_t> radix_bits;
+
+  void swap_input_sides() {
+    std::swap(input_left, input_right);
+    std::swap(data_type_left, data_type_right);
+    std::swap(nullable_left, nullable_right);
+
+    for (auto& secondary_predicate : secondary_predicates) {
+      std::swap(secondary_predicate.column_ids.first, secondary_predicate.column_ids.second);
+      secondary_predicate.predicate_condition = flip_predicate_condition(secondary_predicate.predicate_condition);
+    }
+  }
+
+  auto to_tuple() const {
+    return std::tie(input_left, input_right, join_mode, data_type_left, data_type_right, nullable_left, nullable_right,
+                    predicate_condition, secondary_predicates);
+  }
+};
+
+bool operator<(const JoinTestConfiguration& l, const JoinTestConfiguration& r) { return l.to_tuple() < r.to_tuple(); }
+
+// Virtual interface to create a join operator
+class BaseJoinOperatorFactory {
+ public:
+  virtual ~BaseJoinOperatorFactory() = default;
+  virtual std::shared_ptr<AbstractJoinOperator> create_operator(
+      const std::shared_ptr<const AbstractOperator>& left, const std::shared_ptr<const AbstractOperator>& right,
+      const JoinMode mode, const OperatorJoinPredicate& primary_predicate,
+      const std::vector<OperatorJoinPredicate>& secondary_predicates, const JoinTestConfiguration& configuration) = 0;
+};
+
+template <typename JoinOperator>
+class JoinOperatorFactory : public BaseJoinOperatorFactory {
+ public:
+  std::shared_ptr<AbstractJoinOperator> create_operator(const std::shared_ptr<const AbstractOperator>& left,
+                                                        const std::shared_ptr<const AbstractOperator>& right,
+                                                        const JoinMode mode,
+                                                        const OperatorJoinPredicate& primary_predicate,
+                                                        const std::vector<OperatorJoinPredicate>& secondary_predicates,
+                                                        const JoinTestConfiguration& configuration) override {
```

Comment:
Aren't all of the above part of the configuration?

Reply:
well, at least join_mode and seconary_predicates are direct members of the configuration. All other ones need to be built from info in the configuration.

Reply:
For me, it’s confusing to pass the same parameter twice. What keeps someone from using the `mode` from `configuration`, which might differ from the one passed in?

Reply:
sorry, I wasn't clear. Of course I removed `join mode` and `secondary predicates` from the parameter list and use the one from `configuration`